Sticky Sessions: Tolerate ClickHouse Session ID Mechanism (Better Tests Coverage) + FIPS Support based on BoringSSL

pavelnemirovsky opened the pull request:
Scope of change;
1. Reopened pull request based on https://github.com/ContentSquare/chproxy/pull/119 and continuation of https://github.com/ContentSquare/chproxy/pull/117.
2. Add `make release-build-docker-fips` to be able to build FIPS version of ClickHouse Proxy

pavelnemirovsky committed:
Adjusted logic

pavelnemirovsky committed:
Error handling

pavelnemirovsky committed:
tests

pavelnemirovsky committed:
Adjust

pavelnemirovsky committed:
session_id

pavelnemirovsky committed:
test adjustment

pavelnemirovsky committed:
session_timeout logic

pavelnemirovsky committed:
session_timeout tests

pavelnemirovsky committed:
header test

pavelnemirovsky committed:
cleaning unused code

pavelnemirovsky committed:
handling session_id verification

pavelnemirovsky committed:
Ability to build version under Docker (Native go build doesn't work due to Datadog lib)

pavelnemirovsky committed:
Error handling in Makefile

pavelnemirovsky committed:
Merge https://github.com/Vertamedia/chproxy

pavelnemirovsky committed:
Bump to newer go + make sure static linking works correctly

pavelnemirovsky committed:
Session_id typo fix

pavelnemirovsky committed:
Adjust tests for session-name and session-timeout

pavelnemirovsky committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	Dockerfile
#	Makefile
#	main_test.go
#	proxy.go

pavelnemirovsky committed:
Adjust tests for session-name and session-timeout

pavelnemirovsky commented:
Guys, I am going to add in this PR FIPS 140-2 support based on Borring SSL as well. (Don't review yet)

pavelnemirovsky committed:
Adjust tests for session-name and session-timeout

pavelnemirovsky commented:
@sigua-cs PING

mga-chka commented:
FYI, don't mind the ci error. Your PR is correct and pass the tests, it just that I broke the ci recently to improve our code coverage tracking.
All the maintainers are a bit busy so we might take time to review the PR.

pavelnemirovsky committed:
Adjust tests for session-name and session-timeout

pavelnemirovsky commented:
@mga-chka understood, thx let's try to push forward since the originally provided functionally is not properly covered by tests now.

Blokje5 commented:
@pavelnemirovsky if I understand correctly this is two PRs in one:
1. CH Sticky Sessions
2. Fips support

Would you be able to seperate the PRs? That leads to a cleaner commit history and will make it easier to review.

pavelnemirovsky commented:
@Blokje5 Practically speaking, it is one PR PR since FIPS 140-2 is just a mode in Makefile that allows to compile ClickHouse Proxy with BorringSSL, so I suggest merging it that way.

pavelnemirovsky commented:
Do you want me to add some documentation regarding the ability to build FIPS version of the app?

Blokje5 commented:
@pavelnemirovsky If you would add some documentation regarding the new build that would be great 👍 .

pavelnemirovsky committed:
FIPS and Session_ID Documentation

pavelnemirovsky commented:
@sigua-cs @Blokje5 @mga-chka documentation has been updated. Please check, and let's merge it.

pavelnemirovsky commented:
@sigua-cs @Blokje5 @mga-chka  PING

mga-chka commented:
Sorry for the long delay but we were all off. We'll have a look this week. Can you fix the conflict while we review the PR?

mga-chka commented:
Hi, I took a first look.
The go part regarding the sessionId looks ok.
I'm quite rusty regarding docker so I might be wrong (the last time I touched a docker file was 8 years ago). But, from what I get, there are some problems:
* you're fixing the go version to 1.21 whereas we're setting another value in the [release workflow]( https://github.com/ContentSquare/chproxy/blob/50817d1031ac0f770fb8cae6b0c99cbabb270788/.github/workflows/release.yml#L27) for the moment the version is 1.19, so there is no issue, but the day we need a 1.22+, we'll likely produce failing docker images since there are no tests relating to docker image generation
* the change in the Makefile makes the file Dockerfile not used anymore, why keeping it?
* in the 2 new Dockerfiles, you rely on a `golang:${GO_VER}-alpine` image, which is experimental cf [docker hub](https://hub.docker.com/_/golang) ```This variant is highly experimental, and not officially supported by the Go project (see [golang/go#19938](https://github.com/golang/go/issues/19938) for details).```  whereas the previous image we were using a rock solid debian image
* the new images don't install the packages `ca-certificates` and `curl`, are they already included in the alpine image?

mga-chka commented:
@pavelnemirovsky (I'm pinging you just in case you didn't see my previous msgs)

sconnole commented:
I created a new PR for this issue here: https://github.com/ContentSquare/chproxy/pull/481

**UPDATE**

Pavel gave me write permission, so all changes will be here.

Commit:
chore: update dockerfile, add tests for dockerfile

Added tests for dockerfile to ensure that GoVersions in dockerfiles stay in sync with Go.mod version.
Updated dockerfile to use a more reliable base image.
Install ca-certificates and curl

sconnole commented:
@sigua-cs @Blokje5 @mga-chka  Pinging this on behalf of Pavel

mga-chka commented:
FYI we're in the middle of the holiday in France so most of the maintainer are off, which is why we're not very responsive.
Expect a review in 1-3 weeks (because we will have our yearly contentsquare hackathon once we're not off)

Commit:
chore: remove more vulnerabilities

pavelnemirovsky commented:
@mga-chka I hope your holiday went great; any update?

mga-chka commented:
sorry for the long delay. I took a quick look at the full PR and you didn't take into account my [comment from Jan 4](https://github.com/ContentSquare/chproxy/pull/373#issuecomment-1876098050)

sconnole commented:
@mga-chka 

> sorry for the long delay. I took a quick look at the full PR and you didn't take into account my [comment from Jan 4](https://github.com/ContentSquare/chproxy/pull/373#issuecomment-1876098050)

No stress on the delay. I have added the following changes in the PR that address these comments. 

> you're fixing the go version to 1.21 whereas we're setting another value in the [release workflow](https://github.com/ContentSquare/chproxy/blob/50817d1031ac0f770fb8cae6b0c99cbabb270788/.github/workflows/release.yml#L27) for the moment the version is 1.19, so there is no issue, but the day we need a 1.22+, we'll likely produce failing docker images since there are no tests relating to docker image generation

I added the file `dockerfile_version_test.go` to for testing the dockerfile version is matching the version in `go.mod`. 

> the change in the Makefile makes the file Dockerfile not used anymore, why keeping it?

Updated the dockerfile so the default Dockerfile is used

> in the 2 new Dockerfiles, you rely on a golang:${GO_VER}-alpine image, which is experimental cf [docker hub](https://hub.docker.com/_/golang) This variant is highly experimental, and not officially supported by the Go project (see [golang/go#19938](https://github.com/golang/go/issues/19938) for details). whereas the previous image we were using a rock solid debian image

Changed to using `FROM golang:${GO_VERSION} AS build` instead of the alpine image. I believe the apline image was being used because it's a smaller image. 

> the new images don't install the packages ca-certificates and curl, are they already included in the alpine image?

Added these packages back in.

vsirmbard committed:
Update go.mod

vsirmbard committed:
Update go.sum

pavelnemirovsky committed:
Go 1.22 Fix Action

pavelnemirovsky committed:
Go 1.22 Fix Action

pavelnemirovsky committed:
Merge pull request #1 from pavelnemirovsky/Update-net-and-crypto-packages-version

Update net and crypto packages version

pavelnemirovsky committed:
Merge branch 'master' into master

pavelnemirovsky commented:
@sconnole @mga-chka conflicts were resolved.
